presto-main/src/main/java/io/prestosql/sql/planner/SubqueryPlanner.java

```diff
@@ -108,482 +111,316 @@ public PlanBuilder handleSubqueries(PlanBuilder builder, Collection<Expression>
+                    !analysis.isColumnReference((Expression) expression) && // no point in following dereference chains
+                    !subPlan.canTranslate((Expression) expression)) { // don't consider subqueries under parts of the expression that have already been handled
+                return expression.getChildren();
+            }
 
-    public Set<SubqueryExpression> collectScalarSubqueries(Expression expression, Node node)
-    {
-        return analysis.getScalarSubqueries(node)
-                .stream()
-                .filter(subquery -> nodeContains(expression, subquery))
-                .collect(toImmutableSet());
-    }
+            return ImmutableList.of();
+        };
 
-    public Set<ExistsPredicate> collectExistsSubqueries(Expression expression, Node node)
-    {
-        return analysis.getExistsSubqueries(node)
-                .stream()
-                .filter(subquery -> nodeContains(expression, subquery))
-                .collect(toImmutableSet());
-    }
+        Iterable<Node> allSubExpressions = Traverser.forTree(recurse).depthFirstPreOrder(parent);
 
-    public Set<QuantifiedComparisonExpression> collectQuantifiedComparisonSubqueries(Expression expression, Node node)
-    {
-        return analysis.getQuantifiedComparisonSubqueries(node)
+        return candidates
                 .stream()
-                .filter(quantifiedComparison -> nodeContains(expression, quantifiedComparison.getSubquery()))
-                .collect(toImmutableSet());
+                .filter(candidate -> stream(allSubExpressions).anyMatch(child -> child == candidate))
+                .filter(candidate -> !subPlan.canTranslate(candidate))
+                .collect(toImmutableList());
     }
 
-    private PlanBuilder appendInPredicateApplyNodes(PlanBuilder subPlan, Set<InPredicate> inPredicates, Node node)
+    /**
+     * Group expressions into clusters such that all entries in a cluster are #equals to each other
+     */
+    private <T extends Expression> Collection<List<T>> cluster(List<T> expressions)
     {
-        for (InPredicate inPredicate : inPredicates) {
-            subPlan = appendInPredicateApplyNode(subPlan, inPredicate, node);
+        Map<T, List<T>> sets = new LinkedHashMap<>();
+
+        for (T expression : expressions) {
+            sets.computeIfAbsent(expression, key -> new ArrayList<>())
+                    .add(expression);
         }
-        return subPlan;
+
+        return sets.values();
     }
 
-    private PlanBuilder appendInPredicateApplyNode(PlanBuilder subPlan, InPredicate inPredicate, Node node)
+    private PlanBuilder planInPredicate(PlanBuilder subPlan, List<InPredicate> cluster, Node node)
     {
-        if (subPlan.canTranslate(inPredicate)) {
-            // given subquery is already appended
-            return subPlan;
-        }
+        checkArgument(ImmutableSet.copyOf(cluster).size() == 1, "Cluster contains expressions that don't equal each other");
```

Comment:
I think it makes a sense to extract `Cluster` class

Reply:
What's the benefit? A "cluster" is just a list of expressions, so that would be just a wrapper over a list.

Reply:
>  A "cluster" is just a list of expressions

It is just an implementation of collection of equal elements. Having this abstraction you wouldn't need to call such checks.

Also do you actually need this list? Maybe you could have an representative element and count of copies?